live-build/misc/ansible-roles/appliance-build.dcenter/tasks/main.yml

```diff
@@ -0,0 +1,99 @@
+#
+# Copyright 2018 Delphix
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+---
+- include: ldap.yml
+
+- apt:
+    name: "{{ item }}"
+    state: present
+  with_items:
+    - python-pip
+    - libldap2-dev
+    - libsasl2-dev
+    - isc-dhcp-server
+    - bind9
+    - dnsutils
+    - telnet
+  register: result
+  until: result is not failed
+  retries: 3
+  delay: 60
+
+
+- pip:
+    name: pysphere
+    extra_args: --trusted-host devpi.delphix.com -i http://devpi.delphix.com:3141/root/public/+simple/
+    version: 0.1.8
+
+- pip:
+    name: "{{ item }}"
+    state: present
+  with_items:
+    - paramiko
+    - python-ldap
+    - requests
+  register: result
+  until: result is not failed
+  retries: 3
+  delay: 60
+
+- git:
+    repo: "{{ item.repo }}"
+    dest:
+      "/opt/dcenter/lib/{{ item.dest }}"
+    version: "{{ item.version }}"
+    accept_hostkey: yes
+    update: no
+  with_items:
+    - { repo: 'https://gitlab.delphix.com/devops/dcenter-gate.git',
+        version: projects/dx4linux,
+        dest: dcenter-gate }
+
+- copy:
+    dest: "/etc/dcenter.conf"
```

Comment:
This (and much of the other Ansible code in this role, such as the dependencies on third-party software) seems to me like it more properly belongs in a dedicated Ubuntu package for the DCenter application rather than in the Ansible role. When I first started using Ansible, I recall @tehranian mentioning that "Ansible is not a package manager" and that some jobs are better for Ansible while others are better for package managers. This is just a suggestion, and I don't think it has to be implemented in this change. Perhaps future work could create a proper package for DCenter and move much of this logic into the package management layer.

Reply:
I agree that creating a dcenter package, much like what we do for virtualization and masking, would be a good change.

Reply:
Yup I think once we have infrastructure set-up for hosting misc delphix packages we will be able to convert quite a few things into packages.

Reply:
> once we have infrastructure set-up for hosting misc delphix packages

We already have the infrastructure set up for hosting custom _first-party_ packages (those built from Delphix source repositories). This infrastructure is being used for the Virtualization Engine and Masking Engine packages.

We don't yet have the infrastructure set up for hosting custom _third-party_ packages (those built from non-Delphix source repositories).

Reply:
Right, so ideally I think we'd want to have something more similar to pkg-build-gate for both first and third party packages. Right now we have a separate job and a separate S3 prefix location for each one of the Delphix packages (ZFS, masking, virtualization). This is fine for those larger projects but not very scalable in terms of development effort to add small new packages.

Reply:
That's an interesting idea, but this work isn't currently planned. I'd like to talk more offline to understand the use case you have in mind better.